tests/frontend/io/test_io_xir.py

```diff
@@ -0,0 +1,646 @@
+        assert sf_prog.circuit[1].op.p[0] == FreeParameter("p0")
+        assert sf_prog.circuit[1].op.p[1] == 0.0
+        assert sf_prog.circuit[1].reg[0].ind == 1
+        assert sf_prog.circuit[1].reg[1].ind == 2
+
+        assert sf_prog.circuit[2].op.__class__.__name__ == "Rgate"
+        assert sf_prog.circuit[2].op.p[0] == FreeParameter("p1")
+        assert sf_prog.circuit[2].reg[0].ind == 2
+
+        assert sf_prog.circuit[3].op.__class__.__name__ == "MeasureHomodyne"
+        assert sf_prog.circuit[3].op.p[0] == FreeParameter("p0")
+        assert sf_prog.circuit[3].reg[0].ind == 0
+
+        assert sf_prog.circuit[4].op.__class__.__name__ == "MeasureHomodyne"
+        assert sf_prog.circuit[4].op.p[0] == FreeParameter("p2")
+        assert sf_prog.circuit[4].reg[0].ind == 2
+
+        assert sf_prog.concurr_modes == 5
+        assert sf_prog.timebins == 3
+        assert sf_prog.spatial_modes == 2
+        assert sf_prog.free_params == {
+            "p0": FreeParameter("p0"),
+            "p1": FreeParameter("p1"),
+            "p2": FreeParameter("p2"),
+        }
+        assert all(sf_prog.tdm_params[0] == np.array([np.pi, 3 * np.pi / 2, 0]))
+        assert all(sf_prog.tdm_params[1] == np.array([1, 0.5, np.pi]))
+        assert all(sf_prog.tdm_params[2] == np.array([0, 0, 0]))
+
+
+prog_txt = inspect.cleandoc(
+    """
+    import strawberryfields as sf
+    from strawberryfields import ops
+
+    prog = sf.Program(3)
+    eng = sf.Engine({engine_args})
+
+    with prog.context as q:
+        ops.Sgate(0.54, 0) | q[0]
+        ops.BSgate(0.45, np.pi/2) | (q[0], q[2])
+        ops.Sgate(3*np.pi/2, 0) | q[1]
+        ops.BSgate(2*np.pi, 0.62) | (q[0], q[1])
+        ops.MeasureFock() | q[0]
+
+    results = eng.run(prog)
+    """
+)
+
+prog_txt_no_engine = inspect.cleandoc(
+    """
+    import strawberryfields as sf
+    from strawberryfields import ops
+
+    prog = sf.Program(3)
+
+    with prog.context as q:
+        ops.Sgate(0.54, 0) | q[0]
+        ops.BSgate(0.45, np.pi/2) | (q[0], q[2])
+        ops.Sgate(3*np.pi/2, 0) | q[1]
+        ops.BSgate(2*np.pi, 0.62) | (q[0], q[1])
+        ops.MeasureFock() | q[0]
+"""
+)
+
+prog_txt_tdm = inspect.cleandoc(
+    """
+    import strawberryfields as sf
+    from strawberryfields import ops
+
+    prog = sf.TDMProgram(N=[2, 3])
+    eng = sf.Engine("gaussian")
+
+    with prog.context([np.pi, 3*np.pi/2, 0], [1, 0.5, np.pi], [0, 0, 0]) as (p, q):
+        ops.Sgate(0.123, np.pi/4) | q[2]
+        ops.BSgate(p[0], 0.0) | (q[1], q[2])
+        ops.Rgate(p[1]) | q[2]
+        ops.MeasureHomodyne(p[0]) | q[0]
+        ops.MeasureHomodyne(p[2]) | q[2]
+
+    results = eng.run(prog)
+"""
+)
+
+
+class TestSaveXIR:
+    """Test sf.save functionality"""
+
+    def test_save_filename_path_object(self, prog, tmpdir):
+        """Test saving a program to a file path using a path object"""
+        filename = tmpdir.join("test.xbb")
```

Comment:
```suggestion
        filename = tmpdir.join("test.xir")
```
Same goes for the functions below.